sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,472 @@
+/// - Fees are customizable per Pool.
+/// - Max stored value for both tokens is: U64_MAX / 10_000
+///
+/// This solution is rather simple and is based on the example from the Move repo:
+/// https://github.com/move-language/move/blob/main/language/documentation/examples/experimental/coin-swap/sources/CoinSwap.move
+module defi::pool {
+    use sui::coin::{Self, Coin, TreasuryCap};
+    use sui::balance::{Self, Balance};
+    use sui::id::{VersionedID};
+    use sui::sui::SUI;
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    /// For when supplied Coin is zero.
+    const EZeroAmount: u64 = 0;
+
+    /// For when pool fee is set incorrectly.
+    /// Allowed values are: [0-10000).
+    const EWrongFee: u64 = 1;
+
+    /// For when someone tries to swap in an empty pool.
+    const EReservesEmpty: u64 = 2;
+
+    /// For when initial LSP amount is zero.
+    const EShareEmpty: u64 = 3;
+
+    /// For when someone attemps to add more liquidity than u128 Math allows.
+    const EPoolFull: u64 = 4;
+
+    /// The integer scaling setting for fees calculation.
+    const FEE_SCALING: u128 = 10000;
+
+    /// The max value that can be held in one of the Balances of
+    /// a Pool. U64 MAX / FEE_SCALING
+    const MAX_POOL_VALUE: u64 = {
+        18446744073709551615 / 10000
+    };
+
+    /// The Pool token that will be used to mark the pool share
+    /// of a liquidity provider. The first type parameter stands
+    /// for the witness type of a pool. The seconds is for the
+    /// coin held in the pool.
+    struct LSP<phantom P, phantom T> has drop {}
+
+    /// The pool with exchange.
+    ///
+    /// - `fee_percent` should be in the range: [0-10000), meaning
+    /// that 1000 is 100% and 1 is 0.1%
+    struct Pool<phantom P, phantom T> has key {
+        id: VersionedID,
+        sui: Balance<SUI>,
+        token: Balance<T>,
+        lsp_treasury: TreasuryCap<LSP<P, T>>,
+        fee_percent: u64
+    }
+
+    /// Module initializer is empty - to publish a new Pool one has
+    /// to create a type which will mark LSPs.
+    fun init(_: &mut TxContext) {}
+
+    /// Entrypoint for the `create_pool` method.
+    entry fun create_pool_<P: drop, T>(
+        witness: P,
+        token: Coin<T>,
+        sui: Coin<SUI>,
+        share: u64,
+        fee_percent: u64,
+        ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            create_pool(witness, token, sui, share, fee_percent, ctx),
+            tx_context::sender(ctx)
+        )
+    }
+
+    /// Create new `Pool` for token `T`. Each Pool holds a `Coin<T>`
+    /// and a `Coin<SUI>`. Swaps are available in both directions.
+    ///
+    /// - `share` argument defines the initial amount of LSP tokens
+    /// received by the creator of the Pool.
+    public fun create_pool<P: drop, T>(
+        _: P,
+        token: Coin<T>,
+        sui: Coin<SUI>,
+        share: u64,
+        fee_percent: u64,
```

Comment:
Should say that this is denominated in basis points (easy for someone to say "50" and think that it means 50%).

Reply:
Still think this should say something about basis points.